core/src/main/scala/chisel3/properties/Property.scala

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.properties
+
+import chisel3.{ActualDirection, BaseType, SpecifiedDirection}
+import chisel3.internal.Binding
+import scala.reflect.runtime.universe.{typeOf, TypeTag}
+import scala.annotation.implicitNotFound
+
+/** PropertyType defines a typeclass for valid Property types.
+  *
+  * Typeclass instances will be defined for Scala types that can be used as
+  * properties. This includes builtin Scala types as well as types defined in
+  * Chisel.
+  */
+@implicitNotFound("unsupported Property type ${T}")
+private[chisel3] class PropertyType[T]
+
+/** Companion object for PropertyType.
+  *
+  * Typeclass instances for valid Property types are defined here, so they will
```

Comment:
Will this approach be flexible enough for user-defined struct properties? It seems like everything needs to be declared here to be supported. 
I suppose if the plan is to have some sort of property aggregate base class this would work. Or just only supporting a Scala dictionary for that use case.

Reply:
Yeah it's a good question. I haven't thought about user-defined struct properties yet. We will have some kinds of aggregates, like a Scala Map as a property. Fairly soon, we will also add class/object to Chisel, which is sort of like a Module but only with Property typed ports. And an instance of a class (an object) will itself have a property type. So for user-defined "structs", I think we can support them that way.

In general, I think we will have to have a known set of types that can be used with Properties, if we want to take care of defining how they serialize to FIRRTL. So hopefully the above is enough extensibility, and we can still keep the set of possible Property types a sealed thing.